res/controllers/Pioneer-DDJ-400-script.js

```diff
@@ -0,0 +1,715 @@
+});
+
+PioneerDDJ400.padFxShiftBelowPressed = ignoreRelease(function(channel, control, value, status, group) {
+    var groupAbove = group.replace(/\[EffectRack1_EffectUnit(\d+)_Effect(\d+)]/, function(all, unit, effect) {
+        var effectAbove = parseInt(effect) - 4;
+
+        return "[EffectRack1_EffectUnit" + unit + "_Effect" + effectAbove + "]";
+    });
+
+    engine.setValue(groupAbove, "prev_effect", value);
+});
+
+// END BEAT FX
+
+PioneerDDJ400.vuMeterUpdate = function(value, group) {
+    var newVal = value * 150;
+
+    switch (group) {
+        case "[Channel1]":
+            midi.sendShortMsg(0xB0, 0x02, newVal);
+            break;
+
+        case "[Channel2]":
+            midi.sendShortMsg(0xB1, 0x02, newVal);
+            break;
+    }
+};
+
+//
+// SAMPLERS
+//
+
+// blink pad when sample playback starts
+PioneerDDJ400.samplerPlayOutputCallbackFunction = function(value, group, _control) {
+    if (value === 1) {
+        // for some reason, using script.samplerRegEx here results in an error under Linux
+        var curPad = group.match(/^\[Sampler(\d+)\]$/)[1];
+        startSamplerBlink((0x97 + (curPad > 8 ? 2 : 0)), (0x30 + ((curPad > 8 ? curPad-8 : curPad)-1)), group);
+    }
+};
+
+PioneerDDJ400.samplerModePadPressed = ignoreRelease(function(_channel, control, _value, status, group) {
+    var isLoaded = engine.getValue(group, "track_loaded") === 1;
+
+    if (!isLoaded) {
+        return;
+    }
+    engine.setValue(group, "cue_gotoandplay", 1);
+});
+
+
+PioneerDDJ400.samplerModeShiftPadPressed = function(_channel, _control, value, _status, group) {
+    if (value === 0) {
+        return; // ignore release
+    }
+    var playing = engine.getValue(group, 'play');
+    // when playing stop and return to start/cue point
+    if (playing > 0) {
+        engine.setValue(group, 'cue_gotoandstop', 1);
+    } else { // load selected track
+        // engine.setValue(group, 'LoadSelectedTrack', 1);
+    }
+};
+
+var TimersPioneerDDJ400 = {};
+
+function startSamplerBlink(channel, control, group) {
+    var val = 0x7f;
+
+    // print('channel ' + channel + ' +1= ' + (channel+1));
+
+    stopSamplerBlink(channel, control);
+    TimersPioneerDDJ400[channel][control] = engine.beginTimer(250, function() {
+        val = 0x7f - val;
+
+        // blink the appropriate pad
+        midi.sendShortMsg(channel, control, val);
+        // also blink the pad while SHIFT is pressed
+        midi.sendShortMsg((channel+1), control, val);
+
+        var isPlaying = engine.getValue(group, 'play') === 1;
+
+        if (!isPlaying) {
+            // kill timer
+            stopSamplerBlink(channel, control);
+            // set the pad LED to ON
+            midi.sendShortMsg(channel, control, 0x7f);
+            // set the pad LED to ON while SHIFT is pressed
+            midi.sendShortMsg((channel+1), control, 0x7f);
+        }
+    });
+}
+
+function stopSamplerBlink(channel, control) {
```

Comment:
Please use a function expression instead and make it a member of the `PioneerDDJ400` object.